### PR TITLE
#4 add afterEmit option

### PR DESCRIPTION
**Pull request by daneshka**

### CODE REVIEW

**Comment by happylynx** [comment]
Hi daneshka, thank you for the pull request.

I saw issue #4 you've filed. Could you please elaborate a bit more on why it is beneficial to run flow type checking on a JavaScript bundle?

It seems to be that bundles should have the same interface and behavior as original source code so provided the source code is valid the bundle should also be a valid code. Moreover bundles are stripped of flow type annotations so type-checking them would be less beneficial then checking the annotated source code.

Currently the type checking is run before compilation for plugin to be able stop build process even before expensive compilation phase in case there is a type error.

**Comment by daneshka** [comment]
Hi happylynx and thank you for your response.

Here is how our environment looks like:

- We use Webpack to generate bundles 
- We generate different bundles in the development and production environments. The development bundles are not uglified and are similar to the source code. The only difference is all of the references are merged into a file by webpack.
- We use Flow comments, so we don't strip them in the development bundle.
- We have about 70 libraries. (npm and some none npm)

Why I thought it's good to have it on bundles:
Because it that case, **we don't need to deal with all of the libraries and public types/variables**. Even if we miss an Alias in Webpack config, Flow will let us know.

> Currently the type checking is run before compilation for plugin to be able stop build process even before expensive compilation phase in case there is a type error.

Anytime that a developer changes a line of code and wants to see the result on the web, Webpack regenerate the bundle in a couple of seconds. So by preventing Webpack from doing compilation phase, we don't gain a lot.

**Comment by happylynx** [comment]
Hi daneshka, I tried your patch and it didn't work for me since in time of `after-compile` callback there are no bundle files emitted yet. The `flow` was executed but had nothing to check. How is this supposed to work?

**Comment by daneshka** [comment]
Hi happylynx,
You are right. The event should be `done` instead of `after-compile `
I'll test and update it.

File: src/index.js [hunk 1]
```diff
@@ -63,8 +65,12 @@ FlowWebpackPlugin.prototype.apply = function(compiler) {
         }
     })
 
-    compiler.plugin('run', runCallback(plugin.options.failOnError))
-    compiler.plugin('watch-run', runCallback(plugin.options.failOnErrorWatch))
+    if(plugin.options.bundle) {
+        compiler.plugin('done', runCallback())
```
Comment (happylynx) [inline src/index.js]: I don't find nice to omit an argument of `runCallback()` when the definition expects one. Even though it a valid in JS. Passing `false` would be easier to read.

**Review by happylynx — CHANGES_REQUESTED** [review]
Hi @daneshka, is current version verified?

I suspect that if output of `runCallback()` is registered as `done` hook that is actually sync (unlike async `run` hook) calling `callback()` at line 58 should fail since `callback` will be `undefined`.

File: src/index.js [hunk 2]
```diff
@@ -97,13 +100,17 @@ FlowWebpackPlugin.prototype.apply = function(compiler) {
         compilation.errors.push('Flow validation' + details)
     })
 
+    if(plugin.options.bundle) {
+        compiler.plugin('done', runCallback(false))
+    }else{
     /*
      * callbacks chosen because it is required
      * * to be done before or at time of 'compilation callback' - to avoid expensive compilation when type error present
      * * hook needs to be asynchronous
```
Comment (daneshka) [inline src/index.js:114]: @happylynx  I fixed the issues. But by this comment you mean we shouldn't use 'done' in this case?

Comment (happylynx) [reply src/index.js:114]: Hi @daneshka, sorry for late response. Yes, I believe that `after-emit` phase is best for both running the flow checker and reporting the result (I've added support for user callbacks which required a bit of refactoring). I'd call the option `afterEmit`.

The patch is appreciated, but if you want just checking in your project, you can use this very simplistic checker (`webpack.config.js`):

```js
const flowPath = require.main.require('flow-bin')
const cp = require('child_process')

const flowPlugin = {
    apply: function(compiler) {
        compiler.plugin('after-emit', () => {
            cp.spawnSync(flowPath , [], { stdio: 'inherit' })
        })
    }
}

module.exports = {
    // ...
    plugins: [
        flowPlugin
    ],
}
```

**Pull request closed without merging**